File: Swift/프로그래머스/3/43164. 여행경로/여행경로.swift
import Foundation

var answer: [String] = []
var visited: [String : [Bool]] = [:]
var roadMap: [String : [String]] = [:]

func solution(_ tickets:[[String]]) -> [String] {
    for ticket in tickets {
        var roads = roadMap[ticket[0]] ?? []
        roads.append(ticket[1])
        roadMap[ticket[0]] = roads
    }
    
    for key in roadMap.keys {
        roadMap[key] = roadMap[key]!.sorted(by: { $0 < $1 }) 
        visited[key] = .init(repeating: false, count: roadMap[key]!.count)
    }
    
    dfs("ICN", "ICN", tickets.count, 0)
    
    return answer.first!.split(separator: " ").map { String($0) }
}


func dfs(_ current: String, _ path: String, _ totalTickets: Int, _ count: Int) {
    if (count == totalTickets) {
        answer.append(path)
        return
    }
    
    guard let neighbors = roadMap[current], let _ = visited[current] else { return }
    
    for i in 0..<neighbors.count {
        if !visited[current]![i] {
            visited[current]![i] = true
            let next = neighbors[i]
            
            dfs(next, path + " " + next, totalTickets, count + 1)
            
            visited[current]![i] = false
        }
    }
    
}

File: Java/프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    ArrayList<String> allPaths = new ArrayList<>();
    
    public String[] solution(String[][] tickets) {
        HashMap<String, ArrayList<String>> hm = new HashMap<>();
        
        for (String[] ticket : tickets) {
            ArrayList<String> list = hm.getOrDefault(ticket[0], new ArrayList<>());
            list.add(ticket[1]);
            hm.put(ticket[0], list);
        }
        
        HashMap<String, boolean[]> visited = new HashMap<>();
        for (Map.Entry<String, ArrayList<String>> entry : hm.entrySet()) {
            Collections.sort(entry.getValue());
            visited.put(entry.getKey(), new boolean[entry.getValue().size()]);
        }
        
        dfs("ICN", "ICN", hm, visited, tickets.length, 0);
        
        String[] answer = allPaths.get(0).split(" ");
        return answer;
    }
    
    private void dfs(String current, String path, HashMap<String, ArrayList<String>> hm, HashMap<String, boolean[]> visited, int totalTickets, int count) {
        if (count == totalTickets) {
            allPaths.add(path);
            return;
        }
        
        if (!hm.containsKey(current)) return;
        
        ArrayList<String> neighbors = hm.get(current);
        boolean[] isVisited = visited.get(current);
        
        for (int i = 0; i < neighbors.size(); i++) {
            if (!isVisited[i]) {
                isVisited[i] = true;
                String next = neighbors.get(i);
                
                dfs(next, path + " " + next, hm, visited, totalTickets, count + 1);
                
                isVisited[i] = false;
            }
        }
    }
}